registry.go

```diff
@@ -0,0 +1,114 @@
+package manet
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	ma "github.com/jbenet/go-multiaddr"
+)
+
+type AddrParser func(a net.Addr) (ma.Multiaddr, error)
+type MaddrParser func(ma ma.Multiaddr) (net.Addr, error)
+
+var maddrParsers map[string]MaddrParser
+var addrParsers map[string]AddrParser
+var addrParsersLock sync.Mutex
```

Comment:
woah please no static stuff like this. I know it's common practice in doing it live Go, but this is why we cant have nice things. Users may extend multiaddr with different addrs clashing on a key, and suddenly magically everything could break on them. Static monkey patching like this is pretty bad practice in highly modular code bases. with something like libp2p, we're bound to run into a problem.

this is similar, but actually way more flexible:

```go
type CodecMap struct{
  codecs map[string]NetCodec
  sync.Mutex
}

func (cm *CodecMap) RegisterAddressType(a *AddressSpec) {
  cm.Lock()
  defer cm.Unlock()
  
  for _, k := range a.NetNames {
    cm.codecs[k] = a.NetCodec
  }
  cm.codecs[a.Key] = a.NetCodec
}

var defaultCodecs CodecMap
```